### ISSUE TITLE
Incorporate symlink fix from Chalice upstream

### ISSUE THREAD

**Issue by hannes-ucsc** [DataBiosphere/azul#1921]
https://github.com/aws/chalice/pull/583#issuecomment-648386354

Rebase our fork and remove the cherry picked changes, then release. This would require two PRs, one against https://github.com/DataBiosphere/azul-chalice ad one against Azul.

**Comment by natanlao** [comment]
Clarification from Hannes after standup:
- Only removing symlink-related changes from azul-chalice
- Rebase onto latest stable release from upstream with necessary fixes, unless it is breaking
- Cut release v1.15.1+5

**Comment by natanlao** [comment]
We require https://github.com/aws/chalice/commit/96c670db8b05b6baa07aa672b1acbb3e1ea438ed, which has not yet been released. The latest release at time of writing is v1.15.1.

**Comment by melainalegaspi** [comment]
@hannes-ucsc "1.16.0 has been released and we should be unblocked"

**Comment by natanlao** [comment]
@hannes-ucsc I've pushed my branch [DataBiosphere/azul-chalice@issues/natanlao/1921-symlink-fix](https://github.com/DataBiosphere/azul-chalice/tree/issues/natanlao/1921-symlink-fix). I'm noting this here instead of opening a PR because the my branch is rebased against aws/chalice@1.18.0; there doesn't appear to be anything in DataBiosphere/azul-chalice that I can merge into cleanly.

@jessebrennan I'd appreciate your review on my merge conflict resolution for a commit you authored. My merge is [here](https://github.com/DataBiosphere/azul-chalice/commit/ececd2b17d9c19ecbab0deb13756df6d3cc5dbad) (ececd2b). Here is a patch of the merge conflict:

```diff
diff --cc chalice/package.py
index 5ebcad3,cb473b2..0000000
--- a/chalice/package.py
+++ b/chalice/package.py
@@@ -846,23 -693,22 +846,39 @@@ class TerraformGenerator(TemplateGenera
  
          template['resource'].setdefault('aws_lambda_permission', {})[
              resource.resource_name] = {
++<<<<<<< HEAD
 +            'statement_id': resource.resource_name,
 +            'action': 'lambda:InvokeFunction',
 +            'function_name': resource.lambda_function.function_name,
 +            'principal': self._options.service_principal('s3'),
 +            'source_arn': 'arn:*:s3:::%s' % resource.bucket
++=======
+                 'statement_id': resource.resource_name,
+                 'action': 'lambda:InvokeFunction',
+                 'function_name': self._fref(resource.lambda_function, attr='function_name'),
+                 'principal': 's3.amazonaws.com',
+                 'source_arn': 'arn:aws:s3:::%s' % resource.bucket
++>>>>>>> 76bba2e... Uses TF references for function name
          }
  
      def _generate_sqseventsource(self, resource, template):
          # type: (models.SQSEventSource, Dict[str, Any]) -> None
          template['resource'].setdefault('aws_lambda_event_source_mapping', {})[
              resource.resource_name] = {
++<<<<<<< HEAD
 +            'event_source_arn': self._arnref(
 +                "arn:%(partition)s:sqs:%(region)s"
 +                ":%(account_id)s:%(queue)s",
 +                queue=resource.queue),
 +            'batch_size': resource.batch_size,
 +            'function_name': resource.lambda_function.function_name,
++=======
+                 'event_source_arn': self._arnref(
+                     "arn:aws:sqs:%(region)s:%(account_id)s:%(queue)s",
+                     queue=resource.queue),
+                 'batch_size': resource.batch_size,
+                 'function_name': self._fref(resource.lambda_function, attr='function_name'),
++>>>>>>> 76bba2e... Uses TF references for function name
          }
  
      def _generate_snslambdasubscription(self, resource, template):
@@@ -883,10 -729,10 +899,17 @@@
          }
          template['resource'].setdefault('aws_lambda_permission', {})[
              resource.resource_name] = {
++<<<<<<< HEAD
 +            'function_name': resource.lambda_function.function_name,
 +            'action': 'lambda:InvokeFunction',
 +            'principal': self._options.service_principal('sns'),
 +            'source_arn': topic_arn
++=======
+                 'function_name': self._fref(resource.lambda_function, attr='function_name'),
+                 'action': 'lambda:InvokeFunction',
+                 'principal': 'sns.amazonaws.com',
+                 'source_arn': topic_arn
++>>>>>>> 76bba2e... Uses TF references for function name
          }
  
      def _generate_cloudwatchevent(self, resource, template):
@@@ -924,24 -770,13 +947,32 @@@
          }
          template['resource'].setdefault(
              'aws_lambda_permission', {})[
 +            resource.resource_name] = {
 +            'function_name': resource.lambda_function.function_name,
 +            'action': 'lambda:InvokeFunction',
 +            'principal': self._options.service_principal('events'),
 +            'source_arn': "${aws_cloudwatch_event_rule.%s.arn}" % (
 +                resource.resource_name)
 +        }
 +
 +    def _generate_lambdalayer(self, resource, template):
 +        # type: (models.LambdaLayer, Dict[str, Any]) -> None
 +        template['resource'].setdefault(
 +            "aws_lambda_layer_version", {})[
                  resource.resource_name] = {
++<<<<<<< HEAD
 +                'layer_name': resource.layer_name,
 +                'compatible_runtimes': [resource.runtime],
 +                'filename': resource.deployment_package.filename,
++=======
+                     'function_name': self._fref(resource.lambda_function, attr='function_name'),
+                     'action': 'lambda:InvokeFunction',
+                     'principal': 'events.amazonaws.com',
+                     'source_arn': "${aws_cloudwatch_event_rule.%s.arn}" % (
+                         resource.resource_name)
++>>>>>>> 76bba2e... Uses TF references for function name
          }
 +        self._chalice_layer = resource.resource_name
  
      def _generate_lambdafunction(self, resource, template):
          # type: (models.LambdaFunction, Dict[str, Any]) -> None
@@@ -1039,12 -869,12 +1070,21 @@@
  
          template['resource'].setdefault('aws_lambda_permission', {})[
              resource.resource_name + '_invoke'] = {
++<<<<<<< HEAD
 +            'function_name': resource.lambda_function.function_name,
 +            'action': 'lambda:InvokeFunction',
 +            'principal': self._options.service_principal('apigateway'),
 +            'source_arn':
 +                "${aws_api_gateway_rest_api.%s.execution_arn}/*" % (
 +                    resource.resource_name)
++=======
+                 'function_name': self._fref(resource.lambda_function, attr='function_name'),
+                 'action': 'lambda:InvokeFunction',
+                 'principal': 'apigateway.amazonaws.com',
+                 'source_arn':
+                     "${aws_api_gateway_rest_api.%s.execution_arn}/*" % (
+                         resource.resource_name)
++>>>>>>> 76bba2e... Uses TF references for function name
          }
  
          template.setdefault('output', {})[
```

The part I'm particularly unsure about is `_generate_lambdafunction()`; in my merge, I chose the upstream changes over your original code.

For reference, this is the process that I followed:

- [Comparison](https://github.com/aws/chalice/compare/master...DataBiosphere:hotfix/1.14.0/azul-updates) shows that DataBiosphere/azul-chalice is 12 commits ahead of aws/chalice, five of which appear to be related to symlinks and were removed:
  * [d9f668e](https://github.com/DataBiosphere/azul-chalice/commit/d9f668ed6663f5f9b0d9f844ae53b13465e48895) Added symlinks support for vendor directory walkthrough.
  * [c9a0efa](https://github.com/DataBiosphere/azul-chalice/commit/c9a0efa31dbebfc4edb3a7ec4e9ca9daba2ba855) Fix typecheck.
  * [b3e0228](https://github.com/DataBiosphere/azul-chalice/commit/b3e0228b2d9707240dd1a707f225b870d55344d0) Fix typecheck, again.
  * [6ed0741](https://github.com/DataBiosphere/azul-chalice/commit/6ed0741ad5a336258e472c257780b7f9d6ec1a32) Added tests. 
  * [b17798b](https://github.com/DataBiosphere/azul-chalice/commit/b17798b5cb03fd3ae1c75fbb8a22453064a91d31) Also check symlinks when hashing vendor dir
- `git remote add upstream git@github.com:aws/chalice.git`
- `git fetch --all --prune`
- `git rebase 1.18.0`
- Resolve conflicts
- Update version in setup.py ([51539](https://github.com/DataBiosphere/azul-chalice/commit/51539efe223d1d8ce16df4fba315d5ff6c27201d))

**Comment by jessebrennan** [comment]
My commit was essentially just a find-and-replace of
```
'function_name': resource.lambda_function.function_name
```
with
``` 
'function_name': self._fref(resource.lambda_function, attr='function_name')
```

I recommend excepting all upstream changes, then doing the find-and-replace again afterwords to reconstruct my commit. Let me know if you have questions about that.

Something else caught my eye though which is a bigger concern. It looks like Chalice now automagically supports Lambda Layers. We have our own implementation of this, so upgrading to Chalice 1.18 could cause some problems. We need to investigate first. We should probably get @hannes-ucsc involved in a discussion about this before you proceed.

**Comment by natanlao** [comment]
@jessebrennan 

> I recommend excepting all upstream changes, then doing the find-and-replace again afterwords to reconstruct my commit. Let me know if you have questions about that.

I'll do that, thanks.

> Edit: It looks like the layers stuff is "opt-in" so upgrading to 1.18 should be ok. I started a thread in Slack though where we can discuss it. Anyway you should be good to proceed.

Okay, and I'm assuming that means that in this case I would still be accepting the upstream changes as described above

**Comment by natanlao** [comment]
From standup discussion with Hannes:
1. `git reset --hard` DataBiosphere/azul-chalice@master to aws/chalice@1.18.0
2. Change version from 1.18.0+6 to 1.18.0+5 (do not bump the + as nothing was added)
3. Open a PR against azul-chalice@master
